### PR TITLE
Automatic Dynamic Shapes

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98983
* __->__ #98923

cc @soumith @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -1169,6 +1169,30 @@ def wrap_to_fake_tensor_and_record(
         assert source is not None
         static_shapes, reason = tensor_always_has_static_shape(e, is_tensor)
 
+        name = source.name()
+
+        # Prep for automatic dynamic
```
Comment (voznesenskym) [inline torch/_dynamo/variables/builder.py:1174]: Note - This algorithm could ostensibly be in VariableTracker creation for tensors, invoked through wrap_fx_proxy. This feels cleaner, but open to bikeshedding on where to put it.

**Review by ezyang — APPROVED** [review]
So slick now

**Comment by huydhn** [comment]
@voznesenskym I think this change breaks `dynamic_aot_eager_torchbench` benchmark test for `vision_maskrcnn` with the following error:

```
RuntimeError: Failed running call_method reshape(*(FakeTensor(FakeTensor(..., device='meta', size=(s3, s4, ((s5 + 1)//2), 2)), cuda:0), (s3, s4, s5)), **{}):
maximum recursion depth exceeded in __instancecheck__
```

For example, https://hud.pytorch.org/pytorch/pytorch/commit/ccc9a3d726b1f249daffa442460b70fb50a0ee2e.  Could you help take a look into it?

**Comment by XiaobingSuper** [comment]
@voznesenskym , before this PR, export **TORCHDYNAMO_DYNAMIC_SHAPES=1**, I can get dynamic shape code when running torch benchmark, but after this PR, the generated codes don't have symbolic shapes, is it make sense? or do I need to set another flag? cc @desertfire 

```
TORCHDYNAMO_DYNAMIC_SHAPES=1 python -m torch.backends.xeon.run_cpu --node_id 0 benchmarks/dynamo/timm_models.py --accuracy --float32 -dcpu -n2 --inductor --no-skip --dashboard --only poolformer_m36 --inference
```

**Comment by XiaobingSuper** [comment]
@voznesenskym @desertfire, do you have any comments about https://github.com/pytorch/pytorch/pull/98923#issuecomment-1508225580? do I need to set another flag beside **TORCHDYNAMO_DYNAMIC_SHAPES=1**.  Thanks!

**Comment by desertfire** [comment]
> @voznesenskym @desertfire, do you have any comments about [#98923 (comment)](https://github.com/pytorch/pytorch/pull/98923#issuecomment-1508225580)? do I need to set another flag beside **TORCHDYNAMO_DYNAMIC_SHAPES=1**? Thanks!
> 
> After manually setting `torch._dynamo.config.assume_static_by_default=False`, I can get symbolic shape code(batch size is symbolic), I don't think we need to manually set it when we run model benchmarks. Thanks!

So you are suggesting to control `torch._dynamo.config.assume_static_by_default` with another environment variable? I think auto-dynamic-shapes in this PR is the desired behavior for end users, so I feel control with an explicit flag `torch._dynamo.config.assume_static_by_default` is expected for power users.  cc @voznesenskym @ezyang

**Comment by XiaobingSuper** [comment]
> > @voznesenskym @desertfire, do you have any comments about [#98923 (comment)](https://github.com/pytorch/pytorch/pull/98923#issuecomment-1508225580)? do I need to set another flag beside **TORCHDYNAMO_DYNAMIC_SHAPES=1**? Thanks!
> > After manually setting `torch._dynamo.config.assume_static_by_default=False`, I can get symbolic shape code(batch size is symbolic), I don't think we need to manually set it when we run model benchmarks. Thanks!
> 
> So you are suggesting to control `torch._dynamo.config.assume_static_by_default` with another environment variable? I think auto-dynamic-shapes in this PR is the desired behavior for end users, so I feel control with an explicit flag `torch._dynamo.config.assume_static_by_default` is expected for power users. cc @voznesenskym @ezyang

I think it will be better to have such an environmental variable in case that users know that their model is dynamic. Currently, the strategy seems that generate static shape code at first and then generate dynamic shape code if detects that shape is changed.

I have another question about PyTorch benchmarks test, how do we test our inductor dynamic shape feature if it always gets a static shape code?

File: torch/_dynamo/config.py [hunk 2]
```diff
@@ -55,7 +55,13 @@
...
 # NOTE - this flag can be removed once we can run dynamic_shapes=False w/ the mark_dynamic API
 # see [Note - on the state of mark_dynamic]
-assume_static_by_default = False
+assume_static_by_default = True
+
+# This flag changes how dynamic_shapes=True works, and is meant to be used in conjunction
+# with assume_static_by_default=True.
+# With this flag enabled, we always compile a frame as fully static for the first time, and, if we fail
+# any guards due to wobbles in shape, we recompile with *all* the wobbled shapes as being marked dynamic.
+automatic_dynamic_shapes = True
```
Comment (jgong5) [inline torch/_dynamo/config.py:64]: Just curious: may I know what code uses this configuration flag? I cannot find it or did I miss anything here?

**Comment by ezyang** [comment]
@XiaobingSuper Oops, sorry we missed your comment. To turn off this behavior, toggle `assume_static_by_default`. You are indeed right that `automatic_dynamic_shapes` does nothing, we need to delete it.

**Pull request closed without merging**